### DISCUSSION TITLE
Stable releases

### DISCUSSION THREAD

**Discussion by joelfiddes** [ArcticSnow/TopoPyScale#74] [Ideas]
Ive been having a ton of issues with dependencies recently and also not being able to reproduce old sims because of developments. A lot of this is my own fault probably ;)

It does seem tho there is a generic problem when the environment and code are not bound together:

- if i install a new environment with conda i will get the latest packages which may or may not break things with the latest code, as they can be easily out of sync.
- we need to specify the version of packages that work with the code like in pip "requirements.txt" - can we somehow do that?
- better still can we somehow bundle env, dependencies and code as a release? Then I would only do jobs with a full release and developments in between. As long as you know a release version, the job would be perfectly reproducible without losing track of package versions which quickly becomes a nightmare. Is this what a dockerfile is?

Maybe my workflow is fundamentally :)  flawed but I think something along those lines would be really useful?

**Comment by joelfiddes** [comment]
Maybe this could be cool - so for every code release we also generate  a windos,linux, and mac application:

https://realpython.com/pyinstaller-python/

**Comment by joelfiddes** [comment]
Limitations
PyInstaller is incredibly powerful, but it does have some limitations. Some of the limitations were discussed previously: hidden imports and relative imports in entry-point scripts.

PyInstaller supports making executables for Windows, Linux, and macOS, but it cannot [cross compile](https://en.wikipedia.org/wiki/Cross_compiler). Therefore, you cannot make an executable targeting one Operating System from another Operating System. So, to distribute executables for multiple types of OS, you’ll need a build machine for each supported OS.

Related to the cross compile limitation, it’s useful to know that PyInstaller does not technically bundle absolutely everything your application needs to run. Your executable is still dependent on the users’ [glibc](https://en.wikipedia.org/wiki/GNU_C_Library). Typically, you can work around the glibc limitation by building on the oldest version of each OS you intend to target.

For example, if you want to target a wide array of Linux machines, then you can build on an older version of [CentOS](https://www.centos.org/). This will give you compatibility with most versions newer than the one you build on. This is the same strategy described in [PEP 0513](https://www.python.org/dev/peps/pep-0513/) and is what the [PyPA](https://www.pypa.io/en/latest/) recommends for building compatible wheels.

In fact, you might want to investigate using the [PyPA’s manylinux docker image](https://github.com/pypa/manylinux) for your Linux build environment. You could start with the base image then install PyInstaller along with all your dependencies and have a build image that supports most variants of Linux.

**Comment by ArcticSnow** [comment]
Hey @joelfiddes, I published a new release, v0.2.2, including the `environment.yml` file you started. So now, key packages' versions are specified. Would that be sufficient? The solutions you suggest may be nice from user point of view but increases the maintenance of the package which keeps growing.

**Reply by joelfiddes** [reply]
yup i saw it - nice work! I think this solves the main issue of reproducibility. I need to start storing the version number I run jobs with so can be sure of reproducibility some months later. Maybe somehow writing this to a sim log would do it. We dont produce logs yet actually (another topic!). We can revist the executable stuff later. thanks again.

**Comment by joelfiddes** [comment]
just adding a useful  documentation of reproducible builds here:
https://docs.readthedocs.io/en/stable/guides/reproducible-builds.html
